core/services/directrequest/delegate.go

```diff
@@ -98,22 +121,37 @@ func (l *listener) Start() error {
 	if !connected {
 		return errors.New("Failed to register listener with logBroadcaster")
 	}
-	l.unsubscribeLogs = unsubscribe
+	go l.run()
+
+	hbUnsubscribe := l.headBroadcaster.Subscribe(l)
+
+	l.shutdownWaitGroup.Add(1)
+	go func() {
+		defer hbUnsubscribe()
+		defer unsubscribeLogs()
+		defer l.shutdownWaitGroup.Done()
+		<-l.chStop
+	}()
+
 	return nil
 }
 
 // Close complies with job.Service
 func (l *listener) Close() error {
-	if l.unsubscribeLogs != nil {
-		l.unsubscribeLogs()
+	if !l.OkayToStop() {
+		return errors.New("RegistrySynchronizer is already stopped")
 	}
 	l.runs.Range(func(key, runCloserChannelIf interface{}) bool {
 		runCloserChannel, _ := runCloserChannelIf.(chan struct{})
 		close(runCloserChannel)
 		return true
 	})
 	l.runs = sync.Map{}
+
+	close(l.chStop)
+	l.mbLogs.RetrieveLatestAndClear()
```

Comment:
Does this force l.run to exit before Stop exits?

Reply:
no, i thought i needed it in a test - removed now